pkg/controller/networkpolicy/validate.go

```diff
@@ -0,0 +1,256 @@
+		if err := json.Unmarshal(reqBody, &ar); err != nil {
+			klog.Errorf("CRD validation received incorrect body")
+			admissionResponse = getAdmissionResponseForErr(err)
+		} else {
+			admissionResponse = v.validate(&ar)
+		}
+		aReview := admv1.AdmissionReview{}
+		aReview.TypeMeta.Kind = "AdmissionReview"
+		aReview.TypeMeta.APIVersion = "admission.k8s.io/v1"
+		if admissionResponse != nil {
+			aReview.Response = admissionResponse
+			if ar.Request != nil {
+				aReview.Response.UID = ar.Request.UID
+			}
+		}
+		resp, err := json.Marshal(aReview)
+		if err != nil {
+			klog.Errorf("Unable to encode response during validation: %v", err)
+			http.Error(w, fmt.Sprintf("could not encode response: %v", err), http.StatusInternalServerError)
+		}
+		klog.Infof("Writing validation response to ValidationAdmissionHook")
+		if _, err := w.Write(resp); err != nil {
+			klog.Errorf("Unable to write response during validation: %v", err)
+			http.Error(w, fmt.Sprintf("could not write response: %v", err), http.StatusInternalServerError)
+		}
+	}
+}
+
+// validate function validates a Tier or CNP object
+func (v *NetworkPolicyValidator) validate(ar *admv1.AdmissionReview) *admv1.AdmissionResponse {
+	var result *metav1.Status
+	var msg string
+	allowed := false
+	op := ar.Request.Operation
+	curRaw := ar.Request.Object.Raw
+	oldRaw := ar.Request.OldObject.Raw
+	switch ar.Request.Kind.Kind {
+	case "Tier":
+		klog.Info("Validating Tier CRD")
+		var curTier, oldTier secv1alpha1.Tier
+		if curRaw != nil {
+			if err := json.Unmarshal(curRaw, &curTier); err != nil {
+				klog.Errorf("Error de-serializing current Tier")
+				return getAdmissionResponseForErr(err)
+			}
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldTier); err != nil {
+				klog.Errorf("Error de-serializing old Tier")
+				return getAdmissionResponseForErr(err)
+			}
+		}
+		msg, allowed = v.validateTier(&curTier, &oldTier, op)
+	case "ClusterNetworkPolicy":
+		klog.Info("Validating ClusterNetworkPolicy CRD")
+		var curCNP, oldCNP secv1alpha1.ClusterNetworkPolicy
+		if curRaw != nil {
+			if err := json.Unmarshal(curRaw, &curCNP); err != nil {
+				klog.Errorf("Error de-serializing current ClusterNetworkPolicy")
+				return getAdmissionResponseForErr(err)
+			}
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldCNP); err != nil {
+				klog.Errorf("Error de-serializing old ClusterNetworkPolicy")
+				return getAdmissionResponseForErr(err)
+			}
+		}
+		msg, allowed = v.validateCNP(&curCNP, &oldCNP, op)
+	}
+	if msg != "" {
+		result = &metav1.Status{
+			Message: msg,
+		}
+	}
+	return &admv1.AdmissionResponse{
+		Allowed: allowed,
+		Result:  result,
+	}
+}
+
+// validateCNP validates the admission of a ClusterNetworkPolicy resource
+func (v *NetworkPolicyValidator) validateCNP(curCNP, oldCNP *secv1alpha1.ClusterNetworkPolicy, op admv1.Operation) (string, bool) {
+	allowed := true
+	reason := ""
+	switch op {
+	case "CREATE":
```

Comment:
nit: it seems that K8s provides named constants for these operations: https://godoc.org/k8s.io/kubernetes/pkg/apis/admission#Operation